contract/src/Custody.sol

```diff
@@ -310,54 +293,64 @@ contract Custody is IChannel, IDeposit {
-                uint256 expected = candidate.allocations[CLIENT_IDX].amount + candidate.allocations[SERVER_IDX].amount;
-                if (deposited != expected) {
-                    revert InvalidAllocations();
-                }
+        if (meta.stage == ChannelStatus.INITIAL) {
+            // main goal: verify Candidate == LastValidState, close channel
+            if (!Utils.statesAreEqual(candidate, meta.lastValidState)) {
+                revert InvalidState();
             }
-        }
 
-        if (
-            candidate.data.length == 0
-                || (candidate.intent != StateIntent.INITIALIZE && candidate.intent != StateIntent.RESIZE)
-        ) {
-            // if no state data or intent is not INITIALIZE or RESIZE, assume this is a normal state
+            _closeEffectsAndInteractions(channelId, candidate.allocations);
 
-            // Verify the state is valid according to the adjudicator
-            if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) revert InvalidState();
+            emit Challenged(channelId, block.timestamp);
+            emit Closed(channelId, candidate);
+            return;
+        }
 
-            // Reject states with equal version
-            if (candidate.version == meta.lastValidState.version) {
-                // Explicitly check for equal versions and reject them
-                revert InvalidState();
+        // meta.stage == ChannelStatus.ACTIVE
+        // main goal: verify Candidate is valid and >= LastValidState (in RESIZE case states should be equal)
+        if (lastValidStateIntent == StateIntent.INITIALIZE) {
+            if (candidate.intent == StateIntent.INITIALIZE) {
+                if (!Utils.statesAreEqual(candidate, meta.lastValidState)) revert InvalidState();
+            } else {
+                if (!_isMoreRecent(meta.chan.adjudicator, candidate, meta.lastValidState)) revert InvalidState();
+                if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) {
+                    revert InvalidState();
+                }
             }
-
-            // Revert if trying to challenge with an older state that is already known
-            if (!_isMoreRecent(meta.chan.adjudicator, candidate, meta.lastValidState)) {
-                revert InvalidState();
+        } else if (lastValidStateIntent == StateIntent.OPERATE) {
+            if (candidate.intent != StateIntent.OPERATE) revert InvalidState();
+            if (!Utils.statesAreEqual(candidate, meta.lastValidState)) {
+                if (!_isMoreRecent(meta.chan.adjudicator, candidate, meta.lastValidState)) revert InvalidState();
+                if (!IAdjudicator(meta.chan.adjudicator).adjudicate(meta.chan, candidate, proofs)) {
+                    revert InvalidState();
+                }
+            }
+        } else {
+            // lastValidStateIntent == StateIntent.RESIZE
```

Comment:
Same here, maybe it would be better to specify state intent here

Reply:
8655a27012187f6d247c706367cd5f1a8b804eae